File: packages/nexora_sdk_ios/ios/Classes/HardwareAiManager.swift
import CoreML
import Vision
import Foundation
import MLKitFaceDetection
import MLKitBarcodeScanning
import MLKitTextRecognition
import MLKitVision

class HardwareAiManager {
    var model: VNCoreMLModel?
    
    // ==================== Custom CoreML Model ====================
    
    func loadCustomModel(modelPath: String) -> Bool {
        do {
            let url = URL(fileURLWithPath: modelPath)
            let compiledUrl = try MLModel.compileModel(at: url)
            let mlModel = try MLModel(contentsOf: compiledUrl)
            self.model = try VNCoreMLModel(for: mlModel)
            return true
        } catch {
            return false
        }
    }
    
    func runInference(inputData: Data) -> [String: Any]? {
        let start = Date()
        
        // Generic stub for CoreML inference with raw bytes.
        let executionTime = Date().timeIntervalSince(start) * 1000
        
        return [
            "success": true,
            "executionTimeMs": Int(executionTime)
        ]
    }
    
    // ==================== ML Kit Vision APIs ====================
    
    func processFaceDetection(imageBytes: Data, completion: @escaping ([NexoraAiResult]) -> Void) {
        guard let uiImage = UIImage(data: imageBytes) else {
            completion([])
            return
        }
        let visionImage = VisionImage(image: uiImage)
        visionImage.orientation = uiImage.imageOrientation
        
        let options = FaceDetectorOptions()
        options.performanceMode = .fast
        options.landmarksMode = .all
        options.classificationMode = .all
        
        let faceDetector = FaceDetector.faceDetector(options: options)
        
        faceDetector.process(visionImage) { faces, error in
            guard error == nil, let faces = faces, !faces.isEmpty else {
                completion([])
                return
            }
            
            let results: [NexoraAiResult] = faces.map { face in
                let result = NexoraAiResult()
                result.label = "Face"
                result.confidence = face.hasSmilingProbability ? NSNumber(value: face.smilingProbability) : 1.0
                result.boundingBox = [
                    "top": face.frame.origin.y,
                    "left": face.frame.origin.x,
                    "width": face.frame.size.width,
                    "height": face.frame.size.height
                ]
                return result
            }
            completion(results)
        }
    }
    
    func processBarcodeScanning(imageBytes: Data, completion: @escaping ([NexoraAiResult]) -> Void) {
        guard let uiImage = UIImage(data: imageBytes) else {
            completion([])
            return
        }
        let visionImage = VisionImage(image: uiImage)
        visionImage.orientation = uiImage.imageOrientation
        
        let format = BarcodeFormat.all
        let barcodeOptions = BarcodeScannerOptions(formats: format)
        let barcodeScanner = BarcodeScanner.barcodeScanner(options: barcodeOptions)
        
        barcodeScanner.process(visionImage) { barcodes, error in
            guard error == nil, let barcodes = barcodes, !barcodes.isEmpty else {
                completion([])
                return
            }
            
            let results: [NexoraAiResult] = barcodes.map { barcode in
                let result = NexoraAiResult()
                result.label = "Barcode (\(barcode.format.rawValue))"
                result.confidence = 1.0
                result.recognizedText = barcode.rawValue
                result.boundingBox = [
                    "top": barcode.frame.origin.y,
                    "left": barcode.frame.origin.x,
                    "width": barcode.frame.size.width,
                    "height": barcode.frame.size.height
                ]
                return result
            }
            completion(results)
        }
    }
    
    func processTextRecognition(imageBytes: Data, completion: @escaping ([NexoraAiResult]) -> Void) {
        guard let uiImage = UIImage(data: imageBytes) else {
            completion([])
            return
        }
        let visionImage = VisionImage(image: uiImage)
        visionImage.orientation = uiImage.imageOrientation
        
        let textRecognizer = TextRecognizer.textRecognizer(options: TextRecognizerOptions())
        
        textRecognizer.process(visionImage) { resultText, error in
            guard error == nil, let text = resultText else {
                completion([])
                return
            }
            
            let results: [NexoraAiResult] = text.blocks.map { block in
                let result = NexoraAiResult()
                result.label = "TextBlock"
                result.confidence = 1.0
                result.recognizedText = block.text
                result.boundingBox = [
                    "top": block.frame.origin.y,
                    "left": block.frame.origin.x,
                    "width": block.frame.size.width,
                    "height": block.frame.size.height
                ]
                return result
            }
            completion(results)
        }
    }
}

File: packages/nexora_sdk_android/android/src/main/kotlin/com/nexora/sdk/HardwareAiManager.kt
package com.nexora.sdk

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.nexora.sdk.pigeon.NexoraAiResult
import org.tensorflow.lite.Interpreter
import java.io.File
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel

class HardwareAiManager(private val context: Context) {
    private var interpreter: Interpreter? = null

    // ==================== Custom TF Lite Model ====================

    fun loadCustomModel(modelPath: String): Boolean {
        return try {
            val options = Interpreter.Options().apply {
                setNumThreads(4)
            }
            val file = File(modelPath)
            if (file.exists()) {
                interpreter = Interpreter(file, options)
                true
            } else {
                val assetPath = if (modelPath.startsWith("assets/")) modelPath else "flutter_assets/$modelPath"
                try {
                    val fileDescriptor = context.assets.openFd(assetPath)
                    val inputStream = FileInputStream(fileDescriptor.fileDescriptor)
                    val fileChannel = inputStream.channel
                    val startOffset = fileDescriptor.startOffset
                    val declaredLength = fileDescriptor.declaredLength
                    val buffer = fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
                    interpreter = Interpreter(buffer, options)
                    true
                } catch (assetEx: Exception) {
                    val stream = context.assets.open(assetPath)
                    val bytes = stream.readBytes()
                    val buffer = ByteBuffer.allocateDirect(bytes.size).apply {
                        order(ByteOrder.nativeOrder())
                        put(bytes)
                        rewind()
                    }
                    interpreter = Interpreter(buffer, options)
                    true
                }
            }
        } catch (e: Exception) {
            false
        }
    }

    fun runInference(inputBuffer: ByteBuffer): Map<String, Any>? {
        val interp = interpreter ?: return null
        
        // This is a naive stub for arbitrary dynamic models. In a real TF Lite 
        // production app, you map exact shapes. We'll simulate execution logic.
        val outputBuffer = ByteBuffer.allocateDirect(4) 
        try {
            interp.run(inputBuffer, outputBuffer)
            return mapOf("success" to true)
        } catch (e: Exception) {
            return null
        }
    }

    // ==================== ML Kit Vision APIs ====================

    fun processFaceDetection(imageBytes: ByteArray, callback: (List<NexoraAiResult>) -> Unit) {
        val bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size) ?: run {
            callback(emptyList())
            return
        }
        val image = InputImage.fromBitmap(bitmap, 0)
        
        val options = FaceDetectorOptions.Builder()
            .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
            .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
            .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
            .build()
            
        val detector = FaceDetection.getClient(options)
        
        detector.process(image)
            .addOnSuccessListener { faces ->
                val results = faces.map { face ->
                    NexoraAiResult(
                        label = "Face",
                        confidence = face.smilingProbability?.toDouble() ?: 1.0,
                        boundingBox = mapOf(
                            "top" to face.boundingBox.top.toDouble(),
                            "left" to face.boundingBox.left.toDouble(),
                            "width" to face.boundingBox.width().toDouble(),
                            "height" to face.boundingBox.height().toDouble()
                        ),
                        recognizedText = null
                    )
                }
                callback(results)
            }
            .addOnFailureListener {
                callback(emptyList())
            }
    }

    fun processBarcodeScanning(imageBytes: ByteArray, callback: (List<NexoraAiResult>) -> Unit) {
        val bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size) ?: run {
            callback(emptyList())
            return
        }
        val image = InputImage.fromBitmap(bitmap, 0)
        
        val options = BarcodeScannerOptions.Builder().build() // All formats
        val scanner = BarcodeScanning.getClient(options)
        
        scanner.process(image)
            .addOnSuccessListener { barcodes ->
                val results = barcodes.map { barcode ->
                    NexoraAiResult(
                        label = "Barcode (${barcode.format})",
                        confidence = 1.0,
                        boundingBox = barcode.boundingBox?.let { rect ->
                            mapOf(
                                "top" to rect.top.toDouble(),
                                "left" to rect.left.toDouble(),
                                "width" to rect.width().toDouble(),
                                "height" to rect.height().toDouble()
                            )
                        },
                        recognizedText = barcode.rawValue
                    )
                }
                callback(results)
            }
            .addOnFailureListener {
                callback(emptyList())
            }
    }

    fun processTextRecognition(imageBytes: ByteArray, callback: (List<NexoraAiResult>) -> Unit) {
        val bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size) ?: run {
            callback(emptyList())
            return
        }
        val image = InputImage.fromBitmap(bitmap, 0)
        
        val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
        
        recognizer.process(image)
            .addOnSuccessListener { text ->
                val results = text.textBlocks.map { block ->
                    NexoraAiResult(
                        label = "TextBlock",
                        confidence = 1.0,
                        boundingBox = block.boundingBox?.let { rect ->
                            mapOf(
                                "top" to rect.top.toDouble(),
                                "left" to rect.left.toDouble(),
                                "width" to rect.width().toDouble(),
                                "height" to rect.height().toDouble()
                            )
                        },
                        recognizedText = block.text
                    )
                }
                callback(results)
            }
            .addOnFailureListener {
                callback(emptyList())
            }
    }
}
